amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,565 @@
+
+    device    = abstractproperty()
+    series    = abstractproperty()
+    subseries = abstractproperty()
+    voltage   = abstractproperty()
+    size      = abstractproperty()
+    package   = abstractproperty()
+    speed     = abstractproperty()
+
+    @property
+    def _part(self):
+            return "{}-{}{}{}{}".format(self.series, self.voltage,
+            self.size, self.package, self.speed)
+
+    _dev_osc_mapping = {
+       "GW1N-1" : "OSCH",
+       "GW1N-1P5" : "OSCO",
+       "GW1N-1P5B" : "OSCO",
+       "GW1N-1S" : "OSCH",
+       "GW1N-2" : "OSCO",
+       "GW1N-2B" : "OSCO",
+       "GW1N-4" : "OSC",
+       "GW1N-4B" : "OSC",
+       "GW1N-4C" : "OSC",
+       "GW1N-9" : "OSC",
+       "GW1N-9C" : "OSC",
+       "GW1NR-1" : "OSCH",
+       "GW1NR-2" : "OSCO",
+       "GW1NR-2B" : "OSCO",
+       "GW1NR-4" : "OSC",
+       "GW1NR-4B" : "OSC",
+       "GW1NR-4C" : "OSC",
+       "GW1NR-9" : "OSC",
+       "GW1NR-9C" : "OSC",
+       "GW1NRF-4B" : "OSC",
+       "GW1NS-2" : "OSCF",
+       "GW1NS-2C" : "OSCF",
+       "GW1NS-4" : "OSCZ",
+       "GW1NS-4C" : "OSCZ",
+       "GW1NSE-2C" : "OSCF",
+       "GW1NSER-4C" : "OSCZ",
+       "GW1NSR-2" : "OSCF",
+       "GW1NSR-2C" : "OSCF",
+       "GW1NSR-4" : "OSCZ",
+       "GW1NSR-4C" : "OSCZ",
+       "GW1NZ-1" : "OSCZ",
+       "GW2A-18" : "OSC",
+       "GW2A-18C" : "OSC",
+       "GW2A-55" : "OSC",
+       "GW2A-55C" : "OSC",
+       "GW2AN-18X" : "OSCW",
+       "GW2AN-55C" : "OSC",
+       "GW2AN-9X" : "OSCW",
+       "GW2ANR-18C" : "OSC",
+       "GW2AR-18" : "OSC",
+       "GW2AR-18C" : "OSC"
+    }
+
+    # Each on-chip oscillator has unique properties
+    @property
+    def _osc_props(self):
+        props = {
+            "freq": 0,
+            "div_min": 2,
+            "div_max": 128,
+            "div_step": 2,
+            "div": 0
+        }
+        osc = self._oscillator
+        if osc == "OSC":
+            if "-4" in self.device:
+                props["freq"] = 210000000
+            else:
+                props["freq"] = 250000000
+        elif osc in ["OSCZ", "OSCO"]:
+            props["freq"] = 250000000
+        elif osc in ["OSCF", "OSCH"]:
+            props["freq"] = 240000000
+        elif osc == "OSCW":
+            props["freq"] = 200000000
+        else:
+            assert False
+
+        if not hasattr(self, "osc_frequency"):
+            raise AttributeError('default_clk "OSC" requires attribute "osc_frequency" to be set.')
+
+        # osc_frequency is given in Hz
+        div = props["freq"] // self.osc_frequency
+        props["div"] = div
+
+        if (div < props["div_min"] or div > props["div_max"]
+            or (div % props["div_step"]) != 0):
+                raise ValueError("OSC frequency (osc_frequency) must be chosen such that the oscillator divider is an integer between {} and {} in steps of {}. The calculated value of {!r} is not valid. The undivided oscillator frequency for this device is {} Hz".format(
```

Comment:
Could you split this line to fit a 120 character max length?

Reply:
Fixed.